lib/1.4/utility.dml

```diff
@@ -192,15 +192,105 @@ template _init_val_power_on_reset is (init_val, power_on_reset) {
     }
 }
 
+/**
+### signal_port
+
+<section>
+
+#### Description
+
+Implements a signal interface with saved state. The current state of the signal
+is stored in the saved boolean `high` and an error is logged if the signal is
+raised or lowered when already high or low. The methods `signal_raise` and
+`signal_lower` can be overridden to add additional side effects.
+
+#### Related Templates
+
+poreset, hreset, sreset
+
+</section>
+*/
+template signal_port {
+    implement signal {
+        saved bool high;
+        method signal_raise() default {
+            if (high)
+                log error: "%s raised when already high", qname;
+            else
+                high = true;
+        }
+        method signal_lower() default {
+            if (!high)
+                log error: "%s lowered when already low", qname;
+            else
+                high = false;
+        }
+    }
+}
+
+/**
+### signal_connect
+
+<section>
+
+#### Description
+
+Implements a connect with a signal interface, with saved state. The current
+state of the signal is stored in the saved boolean `signal.high`. If the
+connect is changed while `signal.high` is `true`, the `signal_raise` method
+will be called on the new object. This behaviour can be changed by overriding
+the `set` method. The template defines the following methods, which are default
+and can be overridden:
+
+* `raise()`: Calls `signal_raise` if `signal.high` is false, and sets
+   `signal.high` to `true`.
+
+* `lower()`: Calls `signal_lower` if `signal.high` is true, and sets
+   `signal.high` to `false`.
+
+#### Related Templates
+
+signal_port
+
+</section>
+*/
+template signal_connect is connect {
+    interface signal {
+        saved bool high;
+    }
+
+    method set(conf_object_t *obj) default {
+        /* this is only called if the attribute actually changes */
+        default(obj);
+        if (signal.val && signal.high)
+            signal.signal_raise();
```

Comment:
If instantiating from scratch, this must be delayed (finalization phase or whatever that corresponds to with the new class functions).  If restoring a checkpoint, should not re-raise.

Reply:
More fun: There is no documented guarantee on the attribute setting order between the `signal_high` attribute and this connect on instantiation; in principle this means that the value of `signal.high` is not reliable.

However, to be fair this code is part of the DML distribution, so you can safely rely on the undocumented attribute order, which is alphabetical, so indeed signal.high will be false. This deserves more comments and testing.

Reply:
Still need to test (and comment) the attribute order invariant, i.e. `SIM_create_object(..., [["foo", stub_obj], ["foo_signal_high", True]])` raises the signal.
Looking again at the code, it seems that the init order between attribute/connect/register objects is indeed alphabetical, but looks like all `saved` attributes are initialized after all attributes/connects/registers. Which will cause the test to fail. Maybe it's OK to change init order so `saved` comes first, and rely on that.

Reply:
This is in `post_init` so all attributes, connects, registers, saved's has been set. There's no dependence on order.
```
dml 1.4;
device meh;
import "utility.dml";
connect foo is signal_connect;
```
```
simics> @stub_obj = SIM_create_object('signal-bus', 'signal_bus', [])
simics> @meh = SIM_create_object('meh', 'meh', [['foo', stub_obj], ['foo_signal_high', True]])
simics> @stub_obj.level
1
```

Reply:
Right! sorry for the noise.

Reply:
objects are usually not set as configured until _after_ `post_init`

Reply:
right